### DISCUSSION TITLE
Unable to use `ConnectionFactoryUtils.checkAfterCompletion()` properly

### DISCUSSION THREAD

**Discussion by tbadie** [spring-projects/spring-amqp#2758] [Q&A]
Hi everyone,

In our spring application we are using a transaction to mysql and to rabbitmq. We are using spring boot 3.2.4 and spring amqp 3.1.3.

We rely on the Best Effort One Phase Commit, with a commit to the DB and then Rabbitmq committing. We wanted to make use of the checkAfterCompletion method described here: https://docs.spring.io/spring-amqp/reference/amqp/transactions.html#tx-sync to log when the Rabbitmq transaction fails but the DB one succeeded.

However we're not quite sure where to set that up. We have tried to create a non-transactional receiver which delegates to a transactional one:

```java
public class MessageProcessorWrapper {
     ...

    @RabbitListener(queues = "${messaging.queues.main}")
    public void receiveMessage(Message message) {
        // Aware that this is a global config and doesn't need to be set every time...
        // But for the example sake we're showing it here.
        ConnectionFactoryUtils.enableAfterCompletionFailureCapture(true);

        messageProcessor.receiveMessage(message);

        try {
            ConnectionFactoryUtils.checkAfterCompletion();
        } catch (AfterCompletionFailedException exception) {
            LOGGER.error("The rabbit transaction failed!", exception);
        }
    }
}
```

With the message processor looking like:

```java
public class MessageProcessor {
    @Transactional
    public void receiveMessage(Message message) {
       ....
    }
}
```

The chain of event as I understand it is:

1. org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer#executeListener seems to start the flow (so presumably a rabbit transaction is started?)
2. Our rabbit listener method above is called
3. We call our messageProcessor and a transaction is started
4. The messageProcessor returns and the DB transaction is committed
5. checkAfterCompletion is called but there is nothing to check since the rabbitmq transaction hasn't yet been committed
org.springframework.amqp.rabbit.connection.RabbitUtils#commitIfNecessary is eventually called and may fail without us being able to be notified by it.

There is no obvious place where I can call the checkAfterCompletion after that rabbitmq commit occurs and google doesn't seem to have examples.

Do you have any advice on how to move forward with this please? I hope this is the right place to ask the question.

**Comment by artembilan** [comment]
The transaction is started around `@RabbitListener` POJO method call only if `AbstractMessageListenerContainer` is populated with a `PlatformTransactionManager`.
The logic is like this:
```
	private boolean receiveAndExecute(final BlockingQueueConsumer consumer) throws Exception { // NOSONAR

		PlatformTransactionManager transactionManager = getTransactionManager();
		if (transactionManager != null) {
			try {
				if (this.transactionTemplate == null) {
					this.transactionTemplate =
							new TransactionTemplate(transactionManager, getTransactionAttribute());
				}
				return this.transactionTemplate
						.execute(status -> { // NOSONAR null never returned
							RabbitResourceHolder resourceHolder = ConnectionFactoryUtils.bindResourceToTransaction(
									new RabbitResourceHolder(consumer.getChannel(), false),
									getConnectionFactory(), true);
							// unbound in ResourceHolderSynchronization.beforeCompletion()
							try {
								return doReceiveAndExecute(consumer);
							}
							catch (RuntimeException e1) {
								prepareHolderForRollback(resourceHolder, e1);
								throw e1;
							}
							catch (Exception e2) {
								throw new WrappedTransactionException(e2);
							}
						});
			}
			catch (WrappedTransactionException e) { // NOSONAR exception flow control
				throw (Exception) e.getCause();
			}
		}

		return doReceiveAndExecute(consumer);

	}
```

So, when Rabbit transaction is there, it is indeed wrong to use `ConnectionFactoryUtils.checkAfterCompletion();` inside a `@RabbitListener` method since transaction is going to complete when we return control from that method back to the container.

Looks like the `ConnectionFactoryUtils.checkAfterCompletion();` was designed for `RabbitTemplate` since this is only what I can find in tests:
```
		RabbitTemplate template = new RabbitTemplate(connectionFactory);
		template.setChannelTransacted(true);
		TransactionTemplate tt = new TransactionTemplate(new TestTransactionManager());
		tt.execute(status -> {
			template.convertAndSend("foo", "bar");
			return null;
		});
		assertThat(TransactionSynchronizationManager.hasResource(connectionFactory)).isFalse();
		assertThatIllegalStateException()
			.isThrownBy(() -> (TransactionSynchronizationManager.getSynchronizations()).isEmpty())
			.withMessage("Transaction synchronization is not active");
		assertThatExceptionOfType(AfterCompletionFailedException.class)
			.isThrownBy(() -> ConnectionFactoryUtils.checkAfterCompletion());
```

I think the mentioned logic in the `SimpleMessageListenerContainer` could be improved to call that `ConnectionFactoryUtils.checkAfterCompletion()` in the `finally` block of that `transactionTemplate.execute()`.

As a workaround I only see a way to set `TransactionInterceptor` and some other custom `MethodInterceptor` with a `ConnectionFactoryUtils.checkAfterCompletion()` call for the:
```
	/**
	 * Public setter for the {@link Advice} to apply to listener executions.
	 * <p>
	 * If a {code #setTransactionManager(PlatformTransactionManager) transactionManager} is provided as well, then
	 * separate advice is created for the transaction and applied first in the chain. In that case the advice chain
	 * provided here should not contain a transaction interceptor (otherwise two transactions would be be applied).
	 * @param adviceChain the advice chain to set
	 */
	public void setAdviceChain(Advice... adviceChain) {
```

Does that make sense to you?

**Reply by tbadie** [reply]
Thanks for responding, this is helpful in confirming that there wasn't a straightforward config/option that we had missed.

I have two follow-up questions:
1. How would you see that finally work? Because we probably wouldn't want to let an exception bubble up from the finally directly?
2. I have no experience with MethodInterceptors/TransactionInterceptors so I am struggling a bit to see how to connect the dots to achieve our goal. I understand that they would be a way to add an extra bit of code after the transaction has concluded but I'm not sure how to set it up. I will do some reading on my side, but if you have some extra pointers they would be more than welcomed!

Thanks again!

**Reply by artembilan** [reply]
Looks like that `AfterCompletionFailedException` is going to be emitted with a `ListenerContainerConsumerFailedEvent` after it was bubbled up to the `SimpleMessageListenerContainer.AsyncMessageProcessingConsumer.run()`:
```
			catch (Throwable t) { //NOSONAR
				// by now, it must be an exception
				if (isActive()) {
					logConsumerException(t);
				}
			}
```

Re. `TransactionInterceptor`.

See more in docs: https://docs.spring.io/spring-framework/reference/core/aop-api/advice.html#aop-api-advice-around.

That `TransactionInterceptor` implement this one.
So, you just configure it properly and inject into the mentioned `setAdviceChain()`.

Another one is going to be your custom one as simple as this:
```
	public class CheckAfterCompletionInterceptor implements MethodInterceptor {

		@Nullable
		@Override
		public Object invoke(@NotNull MethodInvocation invocation) throws Throwable {
			try {
				return invocation.proceed();
			}
			finally {
				ConnectionFactoryUtils.checkAfterCompletion();
			}
		}

	}
```

However, this one has to be injected first. So, the proxy is going to call first our custom interceptor, and that `TransactionInterceptor` will be a part of that `invocation.proceed()`.
Therefor a config for listener container is like this: `setAdviceChain(new CheckAfterCompletionInterceptor(), transactionInterceptorBean)`.

Javadocs for `TransactionInterceptor` are pretty good guideline how to configure it.

**Reply by tbadie** [reply]
Ok, based on your answers I've tried to do what I thought would be simpler and was put an `@Around` the method with the `@RabbitListener` annotation to still get access to as much information as possible to make the logging as useful as possible. Turns out that didn't work but it was helpful to play a bit with those AOP concepts. I'll go ahead with the adviceChain approach.

I'm about to go on holiday so I may ask a follow up in a few weeks, but I wanted to thank you again for your answers, they were very helpful in steering me in the right direction!

**Reply by artembilan** [reply]
The `@Around` on the `@RabbitListener` is too late for `ConnectionFactoryUtils.checkAfterCompletion()`.
The problem is that this `@Around` is going to be called *within* transaction from the listener container, but our goal is to check for transaction result when it is done already.
That's why the workaround is not so easy since we have to deal with a little low-level API for that `setAdviceChain()` hook.

Enjoy your time off!

**Reply by tbadie** [reply]
Hi Artem,

After good time off and some other product work I'm back at this and I keep on scratching my head, so I was hoping that you would be able to answer a couple more questions.

I've followed your advice, and I've created a MethodInterceptor (basically the one you suggested with a few adjustments) and I have configured the SimpleRabbitListenerContainerFactory like that:

```java

    // It reuses the default TransactionAttributeSource bean already existing in the project,
    // which might just come from Spring as I don't think that we configure one.
    @Bean(name = "rabbitTransactionInterceptor")
    public TransactionInterceptor rabbitTransactionInterceptor(TransactionAttributeSource transactionAttributeSource,
                                                                                                         PlatformTransactionManager transactionManager) {
        return new TransactionInterceptor((TransactionManager) transactionManager, transactionAttributeSource);
    }

   @Bean
    public RabbitListenerContainerFactory containerFactory(ConnectionFactory connectionFactory,
                                                           StatefulRetryOperationsInterceptor retryOperationsInterceptor,
                                                           @Qualifier("rabbitTransactionInterceptor") TransactionInterceptor transactionInterceptor,
                                                           PlatformTransactionManager transactionManager) {
        // To ensure the exception is appropriately stored.
        ConnectionFactoryUtils.enableAfterCompletionFailureCapture(true);


        SimpleRabbitListenerContainerFactory simpleRabbitListenerContainerFactory = new SimpleRabbitListenerContainerFactory();
        simpleRabbitListenerContainerFactory.setConnectionFactory(connectionFactory);
        simpleRabbitListenerContainerFactory.setChannelTransacted(Boolean.TRUE);
        simpleRabbitListenerContainerFactory.setTransactionManager(transactionManager);
        simpleRabbitListenerContainerFactory.setAdviceChain(new CheckAfterCompletionInterceptor(),
                  transactionInterceptor);

        // Other configuration have been removed for clarity sake.

        return simpleRabbitListenerContainerFactory;
   }
```

Now I've read about it and ran it/stepped through it, and I think that the workaround doesn't really work. But hopefully I'm missing something?

A [stacktrace](https://github.com/user-attachments/files/16751195/InterceptorStackTrace.txt)
starting from within the receiver on the event shows that:

1. The MethodInterceptor (CheckAfterCompletionInterceptor) is called
2. The transaction interceptor as well (a few indirections after)

But while stepping through, it was clear that the TransactionInterceptor was only attaching itself to the already existing transaction that was created before both those Advices were called, meaning that the rabbitmq commit still happens after we check whether an issue occurred, rendering the whole endeavour pointless.

I then tried to find if there was another point in the code that I could use AOP on but didn't really find suitable options given that everything interesting seems to be either final or private. So I have the feeling that at this point the best course of action to get what I want would be to modify the spring library to either make that call later and log or log it when we detect the error. Obviously we'd need a way to let people log if they want/what they want, so probably just a consumer type of interface.

Do you agree with my assessment that the workaround doesn't achieve my objective?
Do you think that adding a configurable consumer for the library users to be able to manage this how they see fit make sense?

Thanks in advance!

**Reply by artembilan** [reply]
Hi @tbadie !

Thank you for sharing your experience!

I think your mistake is here:
```
  simpleRabbitListenerContainerFactory.setTransactionManager(transactionManager);
```
So, setting a transaction manager explicitly into the factory makes that transaction to be started really *before* your advises.

And see my comment before:

>  think the mentioned logic in the `SimpleMessageListenerContainer` could be improved to call that `ConnectionFactoryUtils.checkAfterCompletion()` in the finally block of that `transactionTemplate.execute()`.

So, feel free to open a GH issue on the matter and go ahead with a contribution to make this much more easier and transparent!

**Reply by tbadie** [reply]
Hi @artembilan,

Thanks again for your answer. I indeed see that mistake and I should have spotted that based on the javadoc of the adviceChain method. Sorry about that.

Unfortunately, if I remove the transactionManager, the rabbitmq commit happens in org.springframework.amqp.rabbit.connection.RabbitUtils#commitIfNecessary which is after the advices are executed and I'm back to the same place of checking whether there was an issue before any issue could occur. So I feel that maybe this workaround is just not meant to work for me.

Given that I can't keep bothering you with question until this workaround works, I think it's probably time to try to make a contribution instead. That way more people can benefit from it.

Do we agree that the code change that you meant is this?

```java
    private boolean receiveAndExecute(final BlockingQueueConsumer consumer) throws Exception { // NOSONAR

        PlatformTransactionManager transactionManager = getTransactionManager();
        if (transactionManager != null) {
            try {
                if (this.transactionTemplate == null) {
                    this.transactionTemplate =
                            new TransactionTemplate(transactionManager, getTransactionAttribute());
                }
                return this.transactionTemplate
                        .execute(status -> { // NOSONAR null never returned
                            RabbitResourceHolder resourceHolder = ConnectionFactoryUtils.bindResourceToTransaction(
                                    new RabbitResourceHolder(consumer.getChannel(), false),
                                    getConnectionFactory(), true);
                            // unbound in ResourceHolderSynchronization.beforeCompletion()
                            try {
                                return doReceiveAndExecute(consumer);
                            }
                            catch (RuntimeException e1) {
                                prepareHolderForRollback(resourceHolder, e1);
                                throw e1;
                            }
                            catch (Exception e2) {
                                throw new AbstractMessageListenerContainer.WrappedTransactionException(e2);
                            }
                        });
            }
            catch (AbstractMessageListenerContainer.WrappedTransactionException e) { // NOSONAR exception flow control
                throw (Exception) e.getCause();
            }
            finally {
                ConnectionFactoryUtils.checkAfterCompletion();
            }
        }

        return doReceiveAndExecute(consumer);

    }

```

That way:

1. The check occurs outside of the transaction
2. The exception gets a chance to bubble up and get caught appropriately. 

So all I'd need to do is add a unit test as well. Did I get that right?

**Reply by artembilan** [reply]
Well, let's look if that `simpleRabbitListenerContainerFactory.setChannelTransacted(Boolean.TRUE);` is a cause!
You start transaction yourself in the advice, so no any local transaction for the container then!

Yes, I think your plan sounds good about contribution.

Thank you!

**Reply by tbadie** [reply]
I've removed that line and actually it worked. I couldn't quite believe it so I dug a bit more into it: it worked because the listener I was using broadcast another event to a transactional channel, which meant that this piece of code was called: `org.springframework.amqp.rabbit.connection.ConnectionFactoryUtils#getTransactionalResourceHolder(org.springframework.amqp.rabbit.connection.ConnectionFactory, boolean, boolean)` and in turn meant that the JpaTransactionManager was now aware of the rabbit synchronization. That's what allowed  the proper commit of both transactions (JPA and Rabbit) before reaching my CheckAfterCompletionInterceptor.

So I would have to call this myself within the transaction, but I'm worried about going too much off road with the expectation and making future migrations a nightmare. So I'll give up on the workaround, but thanks for guiding me through the complexity of Spring AMQP and transactions!


I still have one question regarding the code for the contribution. I'd like to actually be able to log some information about the event that has failed. But based on the code that seems impossible: at the level where I'd call the `ConnectionFactoryUtils.checkAfterCompletion()` method I would only have access to the `BlockingQueueConsumer consumer` and the message is only available in the transaction itself [here](https://github.com/spring-projects/spring-amqp/blob/6841624cd08d83c13209269edbf663a03f773c2b/spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/SimpleMessageListenerContainer.java#L1043C4-L1043C11) and it's not a repeatable call. Could you please confirm that?

Thanks!

**Reply by artembilan** [reply]
Well, unfortunately that's what you can put into an exception your throw from your listener method.
This `ConnectionFactoryUtils.checkAfterCompletion()` is exactly for transaction, so we cannot do anything with the message since we have to come back from the consumer to commit it in the `RabbitResourceHolder.commitAll()`.

**Reply by artembilan** [reply]
The mentioned issue is here: https://github.com/spring-projects/spring-amqp/issues/2805
